File: java/Asset/Programs/CreatePrograms.java
/*
   Marketo REST API Sample Code
   Copyright (C) 2016 Marketo, Inc.

   This software may be modified and distributed under the terms
   of the MIT license.  See the LICENSE file for details.
*/
package dev.marketo.samples.Programs;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.io.Reader;
import java.io.Writer;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.NoSuchElementException;
import java.util.Scanner;

import javax.net.ssl.HttpsURLConnection;

import com.eclipsesource.json.JsonArray;
import com.eclipsesource.json.JsonObject;

//the Java sample code on dev.marketo.com uses the minimal-json package
//minimal-json provides easy and fast representations of JSON
//for more information check out https://github.com/ralfstx/minimal-json

public class CreateProgram {
	public String marketoInstance = "https://299-BYM-827.mktorest.com";//Replace this with the host from Admin Web Services
	public String marketoIdURL = marketoInstance + "/identity";	
	public String clientId = "b417d98f-9289-47d1-a61f-db141bf0267f";	//Obtain from your Custom Service in Admin>Launchpoint
	public String clientSecret = "0DipOvz4h2wP1ANeVjlfwMvECJpo0ZYc";	//Obtain from your Custom Service in Admin>Launchpoint
	public String idEndpoint = marketoIdURL + "/oauth/token?grant_type=client_credentials&client_id=" + clientId + "&client_secret=" + clientSecret;
	
	public String name; //name of new Program
	public JsonObject folder; //a JSON input, with 2 required members, id and type
	public String description; //Description for the folder, up to 2,000 characters
	public String type;//program type, "Default", "Event", or "Event with Webinar"
	public String channel;//channel of program, must be available in Channels list of instance
	public JsonArray tags;//optional list of tags, embedded as a JSON Array of JSON objects with members tagType and tagValue;
	public JsonArray costs;//embedded as a JSON Array of JSON objects with members startDate, cost, and note, optional
	
	public static void main(String[] args){
		CreateProgram program = new CreateProgram();
		program.name = "API Test Program";
		program.folder = new JsonObject().add("id", 1035).add("type", "Folder");
		program.description = "Sample API Program";
		program.type = "Default";
		program.channel = "Email Blast";
		
		String result = program.postData();
		System.out.println(result);
	}
	//Make Request
	public String postData(){
		String result = null;
		try {
			String endpoint = marketoInstance + "/rest/asset/v1/programs.json";
			String body = bodyBuilder();
			URL url = new URL(endpoint);
			System.out.println(url);
			
			
			HttpsURLConnection urlConn = (HttpsURLConnection) url.openConnection();
			urlConn.setRequestMethod("POST");
            urlConn.setRequestProperty("accept", "text/json");
            urlConn.setRequestProperty("Content-Type", "application/json");
            urlConn.setRequestProperty("Authorization", "Bearer " + getToken());
            urlConn.setDoOutput(true);
            
            Writer wr = new OutputStreamWriter(urlConn.getOutputStream());
            wr.write(body);
            
			int responseCode = urlConn.getResponseCode();
			if (responseCode == 200){
				InputStream inStream = urlConn.getInputStream();
				result = convertStreamToString(inStream);
			}else{
				result = "Status Code: " + responseCode;
			}
		} catch (MalformedURLException e) {
			e.printStackTrace();
		} catch (IOException e) {
			e.printStackTrace();
		}
		return result;
		
	}
	
	public String bodyBuilder(){
		StringBuilder body = new StringBuilder();
		body.append("name=" + name);
		body.append("&folder=" + folder.toString());
		body.append("&description=" + description);
		body.append("&type=" + type);
		body.append("&channel=" + channel);
		if(tags != null){
			body.append("&tags=" + tags.toString());
		}
		if(costs != null){
			body.append("&costs=" + costs.toString());
		}
		System.out.println(body);
		return body.toString();
	}
	public String getToken(){
		String token = null;
		try {
			URL url = new URL(idEndpoint);
			HttpsURLConnection urlConn = (HttpsURLConnection) url.openConnection();
			urlConn.setRequestMethod("GET");
            urlConn.setRequestProperty("accept", "application/json");
            int responseCode = urlConn.getResponseCode();
            System.out.println(idEndpoint);
            if (responseCode == 200) {
                InputStream inStream = urlConn.getInputStream();
                Reader reader = new InputStreamReader(inStream);
                JsonObject jsonObject = JsonObject.readFrom(reader);
                System.out.println(jsonObject);
                token = jsonObject.get("access_token").asString();
            }else {
                throw new IOException("Status: " + responseCode);
            }
		} catch (MalformedURLException e) {
			e.printStackTrace();
		}catch (IOException e) {
            e.printStackTrace();
        }
		return token;
	}
    private String convertStreamToString(InputStream inputStream) {

        try {
            return new Scanner(inputStream).useDelimiter("\\A").next();
        } catch (NoSuchElementException e) {
            return "";
        }
    }
    
}

File: php/Asset/Programs/CreatePrograms.php
<?php
/*
   CreatePrograms.php

   Marketo REST API Sample Code
   Copyright (C) 2016 Marketo, Inc.

   This software may be modified and distributed under the terms
   of the MIT license.  See the LICENSE file for details.
*/
$program = new CreateProgram();
$program->folder = new stdClass();
$program->folder->id = 5562;
$program->folder->type = "Folder";
$program->name = "New Program PHP";
$program->description = "created with PHP";
$program->type = "Default";
$program->channel = "Content";

print_r($program->postData());


class CreateProgram{
        private $host = "CHANGE ME";
	private $clientId = "CHANGE ME";
	private $clientSecret = "CHANGE ME";

	public $folder;//folders object with id and type
	public $name;//name of new program
	public $description;//description of new program
	public $type;//type of new program
	public $channel;//channel of new Program
	public $tags;//array of tag objects, optional
	public $costs;//array of cost objects, optional
	
	public function postData(){
		$url = $this->host . "/rest/asset/v1/programs.json";
		$ch = curl_init($url);
		$requestBody = $this->bodyBuilder();
		curl_setopt($ch,  CURLOPT_RETURNTRANSFER, 1);
		curl_setopt($ch, CURLOPT_HTTPHEADER, array('accept: application/json', "Authorization: Bearer " . $this->getToken(), "Content-Type: application/json"));
		curl_setopt($ch, CURLOPT_POST, 1);
		curl_setopt($ch, CURLOPT_POSTFIELDS, $requestBody);
		curl_getinfo($ch);
		$response = curl_exec($ch);
		return $response;
	}

	private function getToken(){
		$ch = curl_init($this->host . "/identity/oauth/token?grant_type=client_credentials&client_id=" . $this->clientId . "&client_secret=" . $this->clientSecret);
		curl_setopt($ch,  CURLOPT_RETURNTRANSFER, 1);
		curl_setopt($ch, CURLOPT_HTTPHEADER, array('accept: application/json',));
		$response = json_decode(curl_exec($ch));
		curl_close($ch);
		$token = $response->access_token;
		return $token;
	}
	private function bodyBuilder(){
// 		$requestBody = new stdClass();
// 		$requestBody->folders = $this->folder;
// 		$requestBody->name = $this->name;
// 		$requestBody->description = $this->description;
// 		$requestBody->type = $this->type;
// 		$requestBody->channel = $this->channel;
// 		if (isset($this->tags)){
// 			$requestBody->tags = $this->tags;
// 		}
// 		if(isset($this->costs)){
// 			$requestBody->costs = $this->costs;
// 		}
// 		$json = json_encode($requestBody);
// 		print_r($json);
// 		return $json;
		$jsonFolder = json_encode($this->folder);
		$requestBody = "name=$this->name&folder=$jsonFolder&description=$this->description&type=$this->type&channel=$this->channel";
		if (isset($this->tags)){
			$jsonTags = json_encode($this->tags);
			$requestBody .= "&tags=$jsonTags";
		}
		if (isset($this->costs)){
			$jsonCosts = json_encode($this->costs);
			$requestBody .= "&costs=$jsonCosts";
		}
		return $requestBody;
	}
}